### PR TITLE
admin activity log search: trailing spaces are not trimmed #4512

### LINKED ISSUE #4512: admin activity log search: trailing spaces are not trimmed
Both `info:abc` and `info:abc[space]` should return the same results (irrespective of trailing spaces), but they don't.

### PR DESCRIPTION
**Pull request by RatulGhosh**
Fixes #4512

### CODE REVIEW

File: src/main/java/teammates/common/util/StringHelper.java [hunk 1]
```diff
@@ -507,5 +507,10 @@ public static int countEmptyStrings(String... strings) {
      */
     public static String convertToEmptyStringIfNull(String str) {
         return (str == null) ? "" : str;
-    }  
+    } 
+    public static void trim(String []str){
```
Comment (bamboo3250) [inline src/main/java/teammates/common/util/StringHelper.java]: Could you add a new line between 2 functions?

Comment (bamboo3250) [inline src/main/java/teammates/common/util/StringHelper.java]: And I don't think this will work. Let's make it return a String array as result.

**Comment by bamboo3250** [comment]
And please rename the Pull Request title. It it should be the same as the issue title.
For example: `admin activity log search: trailing spaces are not trimmed #4512`

**Comment by RatulGhosh** [comment]
Sure I will make the necessary changes and rename the PR title.

**Comment by RatulGhosh** [comment]
I made the necessary changes

File: src/main/java/teammates/common/util/StringHelper.java [hunk 2]
```diff
@@ -507,5 +507,12 @@ public static int countEmptyStrings(String... strings) {
      */
     public static String convertToEmptyStringIfNull(String str) {
         return (str == null) ? "" : str;
-    }  
-}
\ No newline at end of file
+    } 
+    
+    public static String[] trim(String []str){
+        for(int i=0;i<str.length;i++){
```
Comment (bamboo3250) [inline src/main/java/teammates/common/util/StringHelper.java]: There is coding standard that we must follow here. Please refer to https://docs.google.com/document/pub?id=1iAESIXM0zSxEa5OY7dFURam_SgLiSMhPQtU0drQagrs&embedded=true
i.e. adding a space between operands and operators, a space after `for` keyword and a space after every `;` and between `)` and `{`.

File: src/main/java/teammates/common/util/StringHelper.java [hunk 3]
```diff
@@ -507,5 +507,12 @@ public static int countEmptyStrings(String... strings) {
...
+            str[i]=str[i].trim();
```
Comment (bamboo3250) [inline src/main/java/teammates/common/util/StringHelper.java]: Likewise for this line. a space between operand and the operator `=`.

**Comment by RatulGhosh** [comment]
@bamboo3250  I have made necessary changes. I will be happy to make additional changes if required. Should I change the function name ?

File: src/main/java/teammates/common/util/StringHelper.java [hunk 4]
```diff
@@ -507,5 +507,12 @@ public static int countEmptyStrings(String... strings) {
      */
     public static String convertToEmptyStringIfNull(String str) {
         return (str == null) ? "" : str;
-    }  
-}
\ No newline at end of file
+    } 
+    
+    public static String[] trim(String []str) {
```
Comment (damithc) [inline src/main/java/teammates/common/util/StringHelper.java]: As this is a public method, I prefer to have a header comment for it.

Comment (damithc) [inline src/main/java/teammates/common/util/StringHelper.java]: `(String[] stringsToTrim)`?

File: src/main/java/teammates/common/util/StringHelper.java [hunk 5]
```diff
@@ -507,5 +507,12 @@ public static int countEmptyStrings(String... strings) {
...
+        for (int i = 0; i < str.length; i++) {
```
Comment (damithc) [inline src/main/java/teammates/common/util/StringHelper.java]: Use a for-each loop?

**Comment by RatulGhosh** [comment]
@damithc  I have made necessary changes .

File: src/main/java/teammates/common/util/StringHelper.java [hunk 6]
```diff
@@ -507,5 +507,18 @@ public static int countEmptyStrings(String... strings) {
...
\ No newline at end of file
+    } 
+    
+    /**
+     * Returns a String array after removing white spaces preceding or
+     * succeeding any string in that array
+     */
+    public static String[] trim(String[] stringsToTrim) {
+        String stringsAfterTrim[] = new String[stringsToTrim.length];
+        int m = 0;
```
Comment (bamboo3250) [inline src/main/java/teammates/common/util/StringHelper.java]: Usually we use `i`, `j` or `k` to iterate through array instead.

File: src/main/java/teammates/common/util/StringHelper.java [hunk 7]
```diff
@@ -507,5 +507,18 @@ public static int countEmptyStrings(String... strings) {
...
+        for (String a : stringsToTrim) {
```
Comment (bamboo3250) [inline src/main/java/teammates/common/util/StringHelper.java]: maybe change `a` to `stringToTrim`?

File: src/main/java/teammates/common/util/StringHelper.java (hunk 6 above)
Comment (RatulGhosh) [inline src/main/java/teammates/common/util/StringHelper.java]: as suggested by @damithc I have used for each loop

**Comment by RatulGhosh** [comment]
@bamboo3250  changed

File: src/main/java/teammates/common/util/StringHelper.java (hunk 6 above)
Comment (bamboo3250) [inline src/main/java/teammates/common/util/StringHelper.java]: I mean the variable `m`. Shouldn't we use `i` or more meaningful variable name instead?

Comment (srajat) [inline src/main/java/teammates/common/util/StringHelper.java]: sorry for misunderstanding . I will make the necessary changes

Comment (RatulGhosh) [inline src/main/java/teammates/common/util/StringHelper.java]: Sorry for the confusion. I will make the change.

**Comment by RatulGhosh** [comment]
@damithc made necessary changes .

File: src/main/java/teammates/common/util/StringHelper.java [hunk 8]
```diff
@@ -507,5 +507,18 @@ public static int countEmptyStrings(String... strings) {
...
-}
\ No newline at end of file
+    } 
+    
+    /**
+     * Returns a String array after removing white spaces preceding or
+     * succeeding any string in that array
+     */
+    public static String[] trim(String[] stringsToTrim) {
+        String stringsAfterTrim[] = new String[stringsToTrim.length];
```
Comment (bamboo3250) [inline src/main/java/teammates/common/util/StringHelper.java]: I think it should be `String[] stringsAfterTrim` here.

File: src/main/java/teammates/common/util/StringHelper.java [hunk 9]
```diff
@@ -507,5 +507,18 @@ public static int countEmptyStrings(String... strings) {
...
     public static String convertToEmptyStringIfNull(String str) {
         return (str == null) ? "" : str;
-    }  
-}
\ No newline at end of file
+    } 
+    
+    /**
+     * Returns a String array after removing white spaces preceding or
+     * succeeding any string in that array
```
Comment (bamboo3250) [inline src/main/java/teammates/common/util/StringHelper.java]: `that array` seems wrong to me. Maybe we can mention the parameter `stringsToTrim` or just `input array`.

**Comment by RatulGhosh** [comment]
Anything else ?

**Comment by bamboo3250** [comment]
Great, it looks good now. :+1:

**Comment by RatulGhosh** [comment]
@bamboo3250 Thanks for your help

File: src/main/java/teammates/common/util/StringHelper.java [hunk 10]
```diff
@@ -507,5 +507,18 @@ public static int countEmptyStrings(String... strings) {
...
     public static String convertToEmptyStringIfNull(String str) {
         return (str == null) ? "" : str;
-    }  
-}
\ No newline at end of file
+    } 
+    
+    /**
+     * Returns a String array after removing white spaces preceding or
+     * succeeding any string in the input array.
```
Comment (damithc) [inline src/main/java/teammates/common/util/StringHelper.java]: `leading and trailing`. 
You can even say `after applying the trim() method of the String class` (as it is a well known method).

**Comment by damithc** [comment]
Small cosmetic change requested.

**Comment by wkurniawan07** [comment]
@RatulGhosh there's a merge conflict here that needs resolve

**Comment by RatulGhosh** [comment]
@wkurniawan07 can you please tell what type of merge conflicts

**Comment by kanghj** [comment]
@RatulGhosh, sync with the main repo and try to merge master into your branch.

**Comment by RatulGhosh** [comment]
please check line no 1991 of [Const.java](https://github.com/TEAMMATES/teammates/blob/master/src/main/java/teammates/common/util/Const.java#L1191) and [src/main/java/teammates/common/util/Const.java  in the comparing changes](https://github.com/TEAMMATES/teammates/compare/master...RatulGhosh:4512-admin-activity-log-search-trailing-spaces-are-not-trimmed).

**Comment by kanghj** [comment]
@RatulGhosh, at the bottom of the pull request page, there is a message that "This branch has conflicts that must be resolved". It also provides a link to a help page on github giving some instructions about resolving merge conflicts. In our development process, under "Fixing Issues" point 7, there are also some instructions about syncing with the committer repo. Could you have a look at that?

**Comment by RatulGhosh** [comment]
Is there still a merge conflict ?

**Comment by kanghj** [comment]
@RatulGhosh, Yes, there is still a merge conflict. Can you undo the last commit where you added the extra line manually, then do a merge commit from master? 

Something like 
<code>
git merge TEAMMATES/master
</code>
 where TEAMMATES is the remote with the url of the main repository

**Comment by wkurniawan07** [comment]
@RatulGhosh, you do not resolve merge conflicts by manually adding in the changes yourself, because as you can see the changes can reach hundreds or thousands of lines. You resolve it by syncing with the target repository (in this case, TEAMMATES' `master` branch), like @kanghj mentioned. Have a look at [this doc](https://git-scm.com/docs/git-merge) to understand how merge works.

**Comment by RatulGhosh** [comment]
I think the merge conflict is solved now . Thanks @wkurniawan07 and @kanghj for help.

File: src/main/java/teammates/common/util/Const.java [hunk 11]
```diff
@@ -1188,7 +1188,6 @@ private void _______values_with_special_meanings________________________(){}
     public static final String GENERAL_QUESTION = "%GENERAL%";
     public static final String USER_IS_TEAM = "%TEAM%";
     public static final String USER_IS_NOBODY = "%NOBODY%";
-    public static final String USER_IS_MISSING = "%MISSING%";
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Const.java]: This line is not supposed to be touched.

File: src/main/java/teammates/common/util/StringHelper.java [hunk 12]
```diff
@@ -528,5 +543,9 @@ public static String removeEnclosingSquareBrackets(String str) {
         }
         
         return str.substring(1, str.length() - 1);
+>>>>>>> master
+=======
+        return stringsAfterTrim;
+>>>>>>> c13ef4e3cda383892bd796acf8a9cce272709070
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/StringHelper.java]: These are the signs of merge conflicts. You should not commit those; rather, you should fix those (see which lines need to be there and which ones don't). If you have not encountered this kind of stuff before, have a look at [this doc](https://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging#Basic-Merge-Conflicts).

**Comment by wkurniawan07** [comment]
@RatulGhosh your `master` branch is not in sync with TEAMMATES' `master` branch. There are extra commits that make it "ahead", and those come from your commits to fix some issue. That is precisely the reason why our workflow prohibits working on a patch in `master` branch. However, do not worry because it can be fixed:

1. On your `master` branch, do a [git hard reset](https://git-scm.com/docs/git-reset) until the point where you have not made committed anything yet. Afterwards sync with TEAMMATES' `master` branch as per what you've done before.
2. Your local git history should now be in sync. To correct your GitHub's `master` branch as well, do a [git force push](https://git-scm.com/docs/git-push).
3. Do the same things to this PR, i.e hard reset until the "changed comment" commit (64b7564), sync with `master` branch (don't forget to fix the merge conflicts correctly), and force push to this branch. That should set things straight.

**Comment by RatulGhosh** [comment]
Done

**Comment by wkurniawan07** [comment]
@RatulGhosh Good job in resolving almost everything! Well, almost, because now as you see you're removing the function `removeEnclosingSquareBrackets` (won't this give a compile error when you try to run the dev server?). It must not be the case, so I'm afraid you have to re-do the step 3 up there, this time make sure not to remove that function.

**Comment by wkurniawan07** [comment]
@RatulGhosh well done :+1:
@bamboo3250 back to you.

**Comment by bamboo3250** [comment]
Thanks, @wkurniawan07. I was so busy recently :sob: I will review this soon.

File: src/main/java/teammates/common/util/StringHelper.java [hunk 13]
```diff
@@ -529,4 +529,17 @@ public static String removeEnclosingSquareBrackets(String str) {
         
         return str.substring(1, str.length() - 1);
     }
+	
+	/**
```
Comment (bamboo3250) [inline src/main/java/teammates/common/util/StringHelper.java]: It seems that you are using tabs here. Could you change them to spaces?

**Comment by thyageshm** [comment]
@bamboo3250 please merge this soon. I am going to be making the release very soon

**Comment by kanghj** [comment]
Merging this now, on behalf of @bamboo3250

**Comment by bamboo3250** [comment]
I couldn't merge this PR on my laptop at the moment.
@kanghj Thanks again :grin:

**Comment by thyageshm** [comment]
thanks @kanghj!

**Pull request merged**